src/main/java/com/example/provider/TripServlet.java

```diff
@@ -208,48 +214,86 @@ public void doPut(HttpServletRequest request, HttpServletResponse response) thro
       sendError(response, "No Trip ID provided.", HttpServletResponse.SC_BAD_REQUEST);
       return;
     }
+
     String tripId = request.getPathInfo().substring(1);
 
     String putData = CharStreams.toString(request.getReader());
     JsonObject jsonBody = GsonProvider.get().fromJson(putData, JsonObject.class);
+
+    logger.info(String.format("Updating trip %s with input %s.", tripId, putData));
+
+    UpdateTripRequest updateRequest = getUpdateTripRequest(jsonBody, tripId, response);
+
+    if (updateRequest == null) {
+      return;
+    }
+
+    TripServiceClient authenticatedServerTripService =
+        grpcServiceProvider.getAuthenticatedTripService();
+
+    Trip updatedTrip = authenticatedServerTripService.updateTrip(updateRequest);
+
+    TripStatus updatedTripStatus = updatedTrip.getTripStatus();
+
+    if (updatedTripStatus == TripStatus.COMPLETE || updatedTripStatus == TripStatus.CANCELED) {
+      logger.info(
+          String.format("Trip %s reached a terminal status. Cleaning up servlet state.", tripId));
+      servletState.setLastTrip(null);
+    }
+
+    logger.info(String.format("Trip:\n%s", updatedTrip));
+    response.getWriter().print(GsonProvider.get().toJson(updatedTrip));
+    response.getWriter().flush();
+  }
+
+  private UpdateTripRequest getUpdateTripRequest(JsonObject jsonBody, String tripId, HttpServletResponse response) throws IOException {
     if (!jsonBody.has(STATUS_INPUT_FIELD)) {
       sendError(response, "No status provided.", HttpServletResponse.SC_BAD_REQUEST);
-      return;
+      return null;
     }
 
     String status = jsonBody.get(STATUS_INPUT_FIELD).getAsString();
+
     TripStatus tripStatus;
+  
     try {
       tripStatus = TripStatus.valueOf(Ascii.toUpperCase(status));
     } catch (IllegalArgumentException e) {
       sendError(response, "Unknown status provided", HttpServletResponse.SC_BAD_REQUEST);
-      return;
+      return null;
     }
 
-    logger.info(String.format("Updating trip %s with status %s.", tripId, status));
+    boolean isUpdatingDestinationIndex = jsonBody.has(INTERMEDIATE_DESTINATION_INDEX_INPUT_FIELD);
```

Comment:
nit: Let's move this variable closer to the `if (isUpdatingDestinationIndex) {` below. Or we can also remove it.